pysph/base/opencl.py

```diff
@@ -35,6 +35,47 @@ def set_queue(q):
     _queue = q
 
 
+class DeviceArray(cl.array.Array):
+    def __init__(self, queue, dtype, n=0):
+        self.queue = queue
+        self.length = n
+        if n == 0:
+            n = 16
+        self.alloc = n
+        self.dtype = dtype
+        self._array = cl.array.empty(queue, n, dtype)
+
+    @property
+    def data(self):
+        return self._array[:self.length]
+
+    def resize(self, size):
+        self.reserve(size)
+        self.length = size
+
+    def reserve(self, size):
+        if size > self.alloc:
+            self._array = cl.array.empty(self.queue, size, self.dtype)
+            self.alloc = size
```

Comment:
The semantics of reserve with STL and other libraries is that it does not destroy existing data.  So you should try to retain whatever old data there already is, in your case you are blowing it away.  See the device helper code...